### PR TITLE
[Feature][RayService]Add kubernetes event to inform user of upgrade strategy

### PR DESCRIPTION
**Pull request by chiayi**
## Why are these changes needed?

Adds Kubernetes Event to inform users of upgrade strategy. Ex:

Screenshot 2024-12-02 at 1 21 01 PM

## Related issue number

Part of https://github.com/ray-project/kuberay/issues/2554

## Checks

- [ ] I've made sure the tests are passing.
- Testing Strategy
   - [ ] Unit tests
   - [x] Manual tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: ray-operator/controllers/ray/rayservice_controller.go [hunk 1]
```diff
@@ -434,12 +434,16 @@ func (r *RayServiceReconciler) reconcileRayCluster(ctx context.Context, rayServi
...
+		strategyMessage := "Neither zero downtime environment var nor UpgradeStrategy is set, enable zero downtime is defaulted to `true`"
 		if zeroDowntimeEnvVar != "" {
 			enableZeroDowntime = strings.ToLower(zeroDowntimeEnvVar) != "false"
+			strategyMessage = fmt.Sprintf("Zero Downtime Environmental Variable is set to `%s`, zero downtime upgrade is set to `%t`", strings.ToLower(zeroDowntimeEnvVar), enableZeroDowntime)
 		}
 		if rayServiceSpecUpgradeStrategy != nil {
 			enableZeroDowntime = *rayServiceSpecUpgradeStrategy == rayv1.NewCluster
+			strategyMessage = fmt.Sprintf("Upgrade Strategy is set to `%s`, zero downtime upgrade is set to `%t`", *rayServiceSpecUpgradeStrategy, enableZeroDowntime)
 		}
+		r.Recorder.Event(rayServiceInstance, "Normal", "UpgradeStrategy", strategyMessage)
```
Comment (andrewsykim) [inline ray-operator/controllers/ray/rayservice_controller.go:446]: I think we previsouly discussed this event only being emitted when the upgrade is triggered. Is this the case if we only emit the event when the cluster action is RolloutNew?

Comment (chiayi) [reply ray-operator/controllers/ray/rayservice_controller.go]: Changed to only emit event if zero downtime triggers an upgrade. 

I've also went back to do another manual test to check the behavior of zero downtime. When there is an upgrade with the "RolloutNew" but with zero downtime disabled, it will continue to retry: 
Screenshot 2024-12-03 at 5 10 28 PM

File: ray-operator/controllers/ray/rayservice_controller.go [hunk 2]
```diff
@@ -435,14 +435,20 @@ func (r *RayServiceReconciler) reconcileRayCluster(ctx context.Context, rayServi
 		// If no fields are set, zero downtime upgrade by default is enabled.
 		// Spec.UpgradeStrategy takes precedence over ENABLE_ZERO_DOWNTIME.
 		enableZeroDowntime := true
+		strategyMessage := ""
 		if zeroDowntimeEnvVar != "" {
 			enableZeroDowntime = strings.ToLower(zeroDowntimeEnvVar) != "false"
+			strategyMessage = fmt.Sprintf("Zero Downtime Environmental Variable is set to `%s`, zero downtime upgrade is set to `%t`", strings.ToLower(zeroDowntimeEnvVar), enableZeroDowntime)
```
Comment (andrewsykim) [inline ray-operator/controllers/ray/rayservice_controller.go:441]: Can you include the full environment variable in this message?
```
ENABLE_ZERO_DOWNTIME environment variable is set to ...
```

**Review by andrewsykim — APPROVED** [review]
small nit

File: ray-operator/controllers/ray/rayservice_controller.go (hunk 2 above)
Comment (chiayi) [reply ray-operator/controllers/ray/rayservice_controller.go]: Done!

File: ray-operator/controllers/ray/rayservice_controller.go [hunk 3]
```diff
@@ -435,14 +435,20 @@ func (r *RayServiceReconciler) reconcileRayCluster(ctx context.Context, rayServi
 		// If no fields are set, zero downtime upgrade by default is enabled.
 		// Spec.UpgradeStrategy takes precedence over ENABLE_ZERO_DOWNTIME.
 		enableZeroDowntime := true
+		strategyMessage := ""
 		if zeroDowntimeEnvVar != "" {
 			enableZeroDowntime = strings.ToLower(zeroDowntimeEnvVar) != "false"
+			strategyMessage = fmt.Sprintf("ENABLE_ZERO_DOWNTIME Environmental Variable is set to `%s`, zero downtime upgrade is set to `%t`", strings.ToLower(zeroDowntimeEnvVar), enableZeroDowntime)
```
[the remark this answers is not in the document]
Comment (chiayi) [reply ray-operator/controllers/ray/rayservice_controller.go]: Done!

File: ray-operator/controllers/ray/rayservice_controller.go [hunk 4]
```diff
@@ -435,14 +435,20 @@ func (r *RayServiceReconciler) reconcileRayCluster(ctx context.Context, rayServi
...
 		}
 		if rayServiceSpecUpgradeStrategy != nil {
 			enableZeroDowntime = *rayServiceSpecUpgradeStrategy == rayv1.NewCluster
+			strategyMessage = fmt.Sprintf("Upgrade Strategy is set to `%s`, zero downtime upgrade is set to `%t`", *rayServiceSpecUpgradeStrategy, enableZeroDowntime)
```
[the remark this answers is not in the document]
Comment (chiayi) [reply ray-operator/controllers/ray/rayservice_controller.go]: Done!

**Review by andrewsykim — APPROVED** [review]
